docs/DafnyRef/Specifications.md

```diff
@@ -601,61 +635,80 @@ interprets JML's `assigns/modifies` in sense (b).
 ACSL and ACSL++ use the `assigns` keyword, but with _modify_ (b) semantics.
 Ada/SPARK's dataflow contracts encode _write_ (a) semantics.
 
-### 5.1.7. Invariant Clause {#sec-invariant-clause}
-````grammar
-InvariantClause_ =
-  "invariant" { Attribute }
-  Expression(allowLemma: false, allowLambda: true)
-````
+### 19.1.7. Invariant Clause ([grammar](#g-invariant-clause)) {#sec-invariant-clause}
+
+Examples:
+<!-- %check-resolve-warn Specifications.2.expect -->
+```dafny
+method m()
+{
+  var i := 10;
+  while 0 < i
+    invariant 0 <= i < 10
+}
+```
 
 An `invariant` clause is used to specify an invariant
 for a loop. If more than one `invariant` clause is given for
 a loop the effective invariant is the conjunction of
-the conditions specified.
+the conditions specified, in their textual order.
 
 The invariant must hold on entry to the loop. And assuming it
-is valid on entry, Dafny must be able to prove that it then
-holds at the end of the loop.
-
-## 5.2. Method Specification {#sec-method-specification}
-````grammar
-MethodSpec =
-  { ModifiesClause(allowLambda: false)
-  | RequiresClause(allowLabel: true)
-  | EnsuresClause(allowLambda: false)
-  | DecreasesClause(allowWildcard: true, allowLambda: false)
+is valid on entry to a particular iteration of the loop, 
+Dafny must be able to prove that it then
+holds at the end of that iteration of the loop.
+
+## 19.2. Method Specification ([grammar](#g-method-specification)) {#sec-method-specification}
+
+Examples:
+<!-- %check-resolve -->
+```dafny
+class C {
+  var next: C?
+  var value: int
+
+  method M(i: int) returns (r: int)
+    requires i >= 0
+    modifies next
+    decreases i
+    ensures r >= 0
+  { 
+    ... 
   }
-````
+}
+```
 
 A method specification is zero or more `modifies`, `requires`,
```

Comment:
```suggestion
A method specification consists of zero or more `modifies`, `requires`,
```

Reply:
This sounds quite _formal_ to me. Same elsewhere. Is this a standard way of expressing these sort of things?

Reply:
I'd say so. This document is a combination of a reference manual (for which formality is typical), a user guide and a tutorial. So the language style is a bit of a mix.